proxy/src/control/discovery.rs

```diff
@@ -154,44 +186,96 @@ impl Discovery {
+                    // wrap the bound service, and insert the store into our map
+                    // so it can be updated later.
+                    let (watch_labels, update_labels) =
+                        futures_watch::Watch::new(meta.metric_labels);
+                    self.metric_labels.insert(addr, update_labels);
+
+                    let service = self.bind.bind(&addr)
+                        .map(|svc| Labeled::new(svc, watch_labels))
+                        .map_err(|_| ())?;
+
+                    return Ok(Async::Ready(Change::Insert(addr, service)))
+                },
+                Update::ChangeMetadata(addr, meta) => match self
+                    .metric_labels
+                    .entry(addr)
+                {
+                    // Handle changes in metadata (currently, destination
+                    // labels by updating our `Store` for that service's set of
+                    // labels.
+                    Entry::Occupied(mut entry) => {
+                        let canceled = entry.get_mut()
+                            .poll_cancel()
+                            .map_err(|e| {
+                                error!("Watch: label poll_cancel: {:?}", e);
+                            });
+                        if canceled?.is_ready() {
+                            // If poll_cancel() returns `Ready`, then the
+                            // service bound to that address has been
+                            // dropped and we can safely remove it.
+                            let _ = entry.remove_entry();
+                            continue;
+                        };
+                        // otherwise, update the store, dropping the previous value.
+                        let _ = entry.get_mut().store(meta.metric_labels)
+                            .map_err(|e| {
+                                error!("Watch: label store errored: {:?}", e);
+                            })?;
+                    },
+                    Entry::Vacant(_) => {
+                        // The store has already been removed, so nobody cares about
+                        // the metadata change.
+                        warn!(
+                            "Watch: ignoring ChangeMetadata for dropped service {:?}",
+                            addr
+                        );
+                    }
+                    // We have nothing to return, so continue with the loop.
+                    // Alternatively, rather than wrapping the method body in
+                    // a loop, we could return `NotReady` here, but we'd have
+                    // to unpark the task. This way, we'll just poll `rx` again,
+                    // (likely returning `NotReady`) and we'll get notified again
+                    // when `rx` is ready...
+                }
+                Update::Remove(addr) => {
+                    return Ok(Async::Ready(Change::Remove(addr)));
```

Comment:
Ah, if we don't do the cancellation, we'd need to remove the item from the metrics_labels here.

Reply:
Yeah, I'll document it.

Reply:
Done in 53c2eb9204e67df1dbf5adbb2ab47436d937f74a.